File: python/logger/logger.py
from typing import List, Union
from .log_types import Handler, ConsoleHandler, ServerHandler, Level, Log, MessageLog, DataLog
import time

class Logger:
    """
    Logger class for python.

    Supports:
        - Server logging (sent to elastic search server)
        - Console logging (prints to console)

    """
    def __init__(self, app_name: str, level: Level = Level.INFO, handlers: List[Handler] = []) -> None:
        if len(app_name.split(" ")) != 1:
            raise ValueError("App name must not contain spaces.")
        self.app_name = app_name
        self.level = level

        # default handler is console handler if not specified
        if not handlers:
            handlers = [ConsoleHandler()]

        # setup handlers
        self.handlers = []
        for handler in handlers:
            if not isinstance(handler, Handler):
                raise ValueError("Handlers must be of type Handler.")
            
            # setup server handler
            if isinstance(handler, ServerHandler):
                handler.setup(self.app_name)

            self.handlers.append(handler)
    
    def _log_message(self, level: Level, message: str):
        if level.value >= self.level.value:
            current_time = time.strftime("%Y-%m-%d %H:%M:%S", time.localtime())
            log: MessageLog = MessageLog(self.app_name, current_time, level, message)
            for handler in self.handlers:
                handler.send(log)
    
    def info(self, message: str):
        """
        Log an information Message Log.

        Args:
            message (str): Message to log.
        """
        self._log_message(Level.INFO, message)
    
    def debug(self, message: str):
        """
        Log a debug Message Log.

        Args:
            message (str): Message to log.
        """
        self._log_message(Level.DEBUG, message)

    def warning(self, message: str):
        self._log_message(Level.WARNING, message)
    
    def error(self, message: str):
        """
        Log an error Message Log.

        Args:
            message (str): Message to log.
        """
        self._log_message(Level.ERROR, message)
    
    def critical(self, message: str):
        """
        Log a critical Message Log.

        Args:
            message (str): Message to log.
        """
        self._log_message(Level.CRITICAL, message)

    def _log_data(self, name: str, value: Union[float, int]):
        current_time = time.strftime("%Y-%m-%d %H:%M:%S", time.localtime())
        log: DataLog = DataLog(self.app_name, current_time, name, value)
        for handler in self.handlers:
            handler.send(log)
    
    def data(self, name: str, value: Union[float, int]):
        """
        Log a data Data Log.

        Args:
            name (str): Name of the data.
            value (Union[float, int]): Value of the data.
        """
        self._log_data(name, value)
    

File: go/logger/logger.go
// logger/logger.go

package logger

import (
	"time"
)

// Logger struct for the logging system
type Logger struct {
	AppName  string
	Level    Level
	Handlers []Handler
}

func NewLogger(appName string, level Level, handlers []Handler) *Logger {
	if len(handlers) == 0 {
		handlers = []Handler{&ConsoleHandler{}}
	}
	for _, handler := range handlers {
		if serverHandler, ok := handler.(*ServerHandler); ok {
			serverHandler.Setup(appName)
		}
	}
	return &Logger{AppName: appName, Level: level, Handlers: handlers}
}

func (l *Logger) logMessage(level Level, message string) {
	if level >= l.Level {
		timestamp := time.Now().Format("2006-01-02 15:04:05")
		log := &MessageLog{AppName: l.AppName, Timestamp: timestamp, Level: level, Message: message}
		for _, handler := range l.Handlers {
			handler.Send(log)
		}
	}
}

func (l *Logger) Info(message string) {
	l.logMessage(INFO, message)
}

func (l *Logger) Debug(message string) {
	l.logMessage(DEBUG, message)
}

func (l *Logger) Warning(message string) {
	l.logMessage(WARNING, message)
}

func (l *Logger) Error(message string) {
	l.logMessage(ERROR, message)
}

func (l *Logger) Critical(message string) {
	l.logMessage(CRITICAL, message)
}

func (l *Logger) LogData(name string, value float64) {
	timestamp := time.Now().Format("2006-01-02 15:04:05")
	log := &DataLog{AppName: l.AppName, Timestamp: timestamp, Name: name, Value: value}
	for _, handler := range l.Handlers {
		handler.Send(log)
	}
}